File: internal/handlers/delete_user_urls.go
package handlers

import (
	"encoding/json"
	"io"
	"net/http"

	"github.com/apetsko/shortugo/internal/models"
)

// DeleteUserURLs handles the deletion of multiple user URLs.
// It accepts a JSON array of short URL IDs in the request body and schedules them for deletion asynchronously.
//
// Request:
//   - Method: DELETE
//   - URL: /api/user/urls
//   - Headers: Content-Type: application/json
//   - Body: ["abc123", "xyz789"]
//
// Response:
//   - 202 Accepted: The deletion request is accepted for processing.
//   - 400 Bad Request: Invalid request body or JSON format.
//   - 401 Unauthorized: User authentication failed.
//
// The function retrieves the user ID from a cookie, validates the request body, and then
// sends the batch delete request to the processing channel asynchronously.
func (h *URLHandler) DeleteUserURLs(w http.ResponseWriter, r *http.Request) {
	// Retrieve the user ID from the cookie
	userID, err := h.Auth.CookieGetUserID(r, h.Secret)
	if err != nil {
		h.Logger.Error(err.Error())
		w.WriteHeader(http.StatusUnauthorized)
		return
	}

	// Read the request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		h.Logger.Info("Error unmarshaling request body", "error", err.Error())
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	var ids []string

	// Unmarshal the JSON array of short URL IDs
	err = json.Unmarshal(body, &ids)
	if err != nil {
		h.Logger.Info("Error unmarshaling request body", "error", err.Error())
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	// Send the batch delete request asynchronously
	go func() {
		h.ToDelete <- models.BatchDeleteRequest{Ids: ids, UserID: userID}
	}()

	// Respond with 202 Accepted and return the requested IDs
	w.WriteHeader(http.StatusAccepted)
	if err := json.NewEncoder(w).Encode(ids); err != nil {
		h.Logger.Error(err.Error())
	}
}

File: wrk/delete_user_urls.lua
wrk.method = "POST"
local headers = {}
headers["Content-Type"] = "application/json"
headers["Cookie"] =
"shortugo=MTc0Mzg3NjM3NXwxc254dHo2b3NRd2t2OXB3R1JGeVUzUGt5Mk80Z2F3ZWgyVVlsclZKVmtzRmhZOHl8DmVfNXB015eCtBY4Rmf2elFRLKy1hYV5mblMS6JmJ0I=; Path=/; HttpOnly;"

local counter = 1400000

function request()
    n1 = counter + 1
    n2 = counter + 2
    n3 = counter + 3
    counter = counter + 3
    local body = string.format('["%d","%d","%d"]', n1, n2, n3)
    return wrk.format("DELETE", "/api/user/urls", headers, body)
end
